### DISCUSSION TITLE
(std::min)(x, y)

### DISCUSSION THREAD

**Discussion by matta** [abseil/abseil-cpp#1271] [Q&A] [answered]
I notice that often Abseil calls `std::min` and `std::max` like this:

```c++
capacity = (std::min)(capacity, kCustomLimit);
```
Why?

See https://github.com/search?l=&p=1&q=std%3A%3Amin+repo%3Aabseil%2Fabseil-cpp+extension%3Ah&ref=advsearch&type=Code

**Comment by derekmauro** [comment, marked as answer]
We use `(std::min)(a, b)` in headers because Windows has macros named `min` and `max`. Our own build rules have [`/DNOMINMAX`](https://github.com/abseil/abseil-cpp/blob/665ac5b425cea2462bf85dcbb56dd68ec19ad00f/absl/copts/copts.py#L41), but we can't guarantee that everyone `#include`-ing our headers also has `/DNOMINMAX`. Somehow the parenthesis is a workaround, although I don't know what parsing rule makes it work.

https://godbolt.org/z/M8Grc6xab

**Reply by matta** [reply]
Ah yeah, with "function style" macros I bet the preprocessor stage needs to see a '(' token after the macro's name, and the parenthesis defeats that.
